packages/compass-aggregations/src/modules/pipeline-builder/stage-editor.ts

```diff
@@ -453,28 +542,57 @@ export const addStage = (
   after?: number
 ): PipelineBuilderThunkAction<void, StageAddAction> => {
   return (dispatch, getState, { pipelineBuilder }) => {
-    const stage = pipelineBuilder.addStage(after);
+    const {
+      pipelineBuilder: {
+        stageEditor: { stages },
+      },
+    } = getState();
+    const addAfter = after ?? stages.length;
+    const addAfterIdxInPipeline = storeIndexToPipelineIndex(stages, addAfter, {
+      // The item at afterIndex could be a wizard hence we need to consider
+      // that while calculating the corresponding afterIdx in pipeline
+      includeIndex: true,
+    });
+
+    const stage = pipelineBuilder.addStage(addAfterIdxInPipeline);
     track('Aggregation Edited', {
-      num_stages: getState().pipelineBuilder.stageEditor.stages.length,
+      num_stages: pipelineFromStore(stages).length,
       stage_action: 'stage_added',
       stage_index: stage.id + 1,
       editor_view_type: mapPipelineModeToEditorViewType(getState()),
     });
-    dispatch({ type: StageEditorActionTypes.StageAdded, after, stage });
+    dispatch({
+      type: StageEditorActionTypes.StageAdded,
+      after: addAfter,
+      afterIdxInPipeline: addAfterIdxInPipeline,
+      stage,
+    });
   };
 };
 
 export const removeStage = (
   at: number
 ): PipelineBuilderThunkAction<void, StageRemoveAction> => {
   return (dispatch, getState, { pipelineBuilder }) => {
-    const num_stages = getState().pipelineBuilder.stageEditor.stages.length;
-    const stage = pipelineBuilder.removeStage(at);
+    const {
+      pipelineBuilder: {
+        stageEditor: { stages },
+      },
+    } = getState();
+    const stageInStore = stages[at];
+    if (!stageInStore || stageInStore.type !== 'stage') {
+      return;
+    }
+
+    const { idxInPipeline } = stageInStore;
+
+    pipelineBuilder.cancelPreviewForStage(idxInPipeline);
+    const stage = pipelineBuilder.removeStage(idxInPipeline);
     track('Aggregation Edited', {
-      num_stages,
+      num_stages: pipelineFromStore(stages).length,
       stage_action: 'stage_deleted',
       stage_name: stage.operator,
-      stage_index: at + 1,
+      stage_index: idxInPipeline + 1,
```

Comment:
I have a feeling that this is incorrect and was incorrect in old implementation as well. Could you help me confirm this @mabaasit ?

The `idxInPipeline` is already the correct index of item to be removed, we don't need to do `+1` to it.

Reply:
It seems to be correct in the old and new implementation. We are adding +1 in every event i think.